File: lab-redes/python/multicast/cliente_multicast.py
import socket
import struct

OFFSET = 22

GRUPO_MULTICAST = "230.0.0.1"
PORTA = 4446 + OFFSET

sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
sock.bind(("", PORTA))

grupo = socket.inet_aton(GRUPO_MULTICAST)
solicitacao_membro = struct.pack("4sL", grupo, socket.INADDR_ANY)
sock.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, solicitacao_membro)

print(f"[Multicast] Inscrito no grupo {GRUPO_MULTICAST}:{PORTA}. Aguardando avisos...")
while True:
    dados, endereco = sock.recvfrom(1024)
    print(f"[Multicast] Recebido: {dados.decode('utf-8')}")

File: lab-redes/java/multicast/ClienteMulticast.java
import java.net.*;
import java.io.IOException;

public class ClienteMulticast {

    static final int OFFSET = 22;

    public static void main(String[] args) throws IOException {
        String grupoMulticast = "230.0.0.1";
        int porta = 4446 + OFFSET;

        try (MulticastSocket socket = new MulticastSocket(porta)) {
            InetAddress grupo = InetAddress.getByName(grupoMulticast);
            InetSocketAddress endpointGrupo = new InetSocketAddress(grupo, porta);

            // Em Wi-Fi corporativa/VPN o multicast costuma ser bloqueado. Para testar
            // servidor e cliente na MESMA máquina, use a interface de loopback explicitamente:
            // NetworkInterface interfaceRede = NetworkInterface.getByInetAddress(InetAddress.getLoopbackAddress());
            NetworkInterface interfaceRede = NetworkInterface.getByInetAddress(InetAddress.getLocalHost());

            socket.joinGroup(endpointGrupo, interfaceRede);
            System.out.println("[Multicast] Inscrito no grupo " + grupoMulticast + ":" + porta + ". Aguardando avisos...");

            byte[] buffer = new byte[1024];
            while (true) {
                DatagramPacket pacote = new DatagramPacket(buffer, buffer.length);
                socket.receive(pacote);
                String mensagem = new String(pacote.getData(), 0, pacote.getLength());
                System.out.println("[Multicast] Recebido: " + mensagem);
            }
        }
    }
}